python/ciqueue/distributed.py

```diff
@@ -0,0 +1,148 @@
+import os
+import time
+import math
+from redis import ConnectionError
+
+from ciqueue.static import Static
+
+
+class LostMaster(StandardError):
+  pass
+
+
+class Base(object):
+  def __init__(self, redis, build_id):
+    self.redis = redis
+    self.build_id = str(build_id)
+    self.is_master = False
+    self._scripts = {}
+
+  def key(self, *args):
+    return ':'.join(['build', self.build_id] + map(str, args))
+
+  def wait_for_master(self, timeout=10):
+    if self.is_master:
+      return True
+
+    for i in range(timeout * 10 + 1):
+      master_status = self._master_status()
+      if master_status in ['ready', 'finished']:
+        return True
+      time.sleep(0.1)
+
+    raise LostMaster("The master worker is still `" + repr(master_status) + "` after 10 seconds waiting.")
+
+  def _master_status(self):
+    return self.redis.get(self.key('master-status'))
+
+  def __len__(self):
+    transaction = self.redis.pipeline(transaction=True)
+    transaction.llen(self.key('queue'))
+    transaction.zcard(self.key('running'))
+    return sum(transaction.execute())
+
+  @property
+  def progress(self):
+    return self.total - len(self)
+
+
+class Worker(Base):
+  def __init__(self, tests, worker_id, redis, build_id, timeout, max_requeues=0, requeue_tolerance=0):
+    super(Worker, self).__init__(redis=redis, build_id=build_id)
+    self.timeout = timeout
+    self.total = len(tests)
+    self.max_requeues = max_requeues
+    self.global_max_requeues = math.ceil(len(tests) * requeue_tolerance)
```

Comment:
It's for the requeue feature. The `pytest` plugin doesn't support it yet (and probably won't for the first version). 

The idea is that you can allow tests to be put back in the queue if they failed, so that if it's a leaky test it will have better chances to succeed.

This value sets a global % of tests that can be requeued. E.g. you have 200 tests, you set this to `0.1` to allow up to 20 requeues per builds.